### PR TITLE
test: property/invariant tests for writable_paths jail surface (#282)

### LINKED ISSUE #282: Property/invariant tests for writable_paths public surface (follow-up to #275)
Follow-up to #275. The #275 review process surfaced ~30 distinct issues across 13 rounds. Most were on first-principles properties of the public surface (path containment, glob matching, error sentinels, syscall flag combinations) that should have been pinned by tests at the implementation moment, not discovered round-by-round under review.

## Scope

Write invariant-style tests for each public function in the jail's surface so a regression in any documented property fails fast with a meaningful diff:

- `tracker.exec.ValidateWritablePaths` — each refusal class (escape vs malformed), each sentinel-wrap contract, each happy doublestar shape. (Partial coverage landed in commit `88463c6`; expand to property-based with random inputs.)
- `tracker.exec.OpenForWrite` — RESOLVE flag enforcement, EACCES vs EXDEV/ELOOP classification, parent symlink rejection, absolute-path refusal, short-path edge cases.
- `tracker.exec.SafeMkdirAll` + `SafeRemove` — symlink rejection at any depth, EEXIST race tolerance, NO_MAGICLINKS coverage, anchor escape via component substitution.
- `pipeline.handlers.configureJail` — every G1/G2/G3 refusal path, env-hook installation, anchor resolution against arbitrary session roots.
- `pipeline.handlers.refuseWritablePathsOnUnsupportedBackend` — happy / acp / claude-code / unknown / nil-node × writable-vs-unset.
- WriteOpener / Remover closures inside configureJail — glob match acceptance/refusal, ErrPathNotAllowed vs ErrPathEscape error class, mkdir ordering.

## Style

Prefer property assertions over example assertions:

```go
// example test
got := matchOneGlob("workspace/foo.md", "workspace/**")
want := true

// property test
quick.Check(func(prefix string, leaf string) bool {
    if !validLiteralSegment(prefix) || !validLiteralSegment(leaf) {
        return true // skip invalid inputs
    }
    g := prefix + "/**"
    return matchOneGlob(prefix+"/"+leaf, g) == true
}, nil)
```

`testing/quick` is in stdlib; no new deps.

## Why

Linux security primitives have decades of edge cases (errno classifications, kernel-version-dependent semantics, syscall flag combinations). Examples in #275: EACCES vs escape, `PR_SET_PDEATHSIG` thread vs process scoping, `openat2` resolve flag combos, Landlock directory-prefix vs glob mismatch. Property tests force the implementer to articulate the contract precisely; example tests just demonstrate it on hand-picked inputs.

## Estimate

~200-400 LOC depending on how thorough. Worth it: each round of #275 review took ~20-40 min to address; this prevents ~half of them next time.

Refs #275.

### PR DESCRIPTION
**Pull request by clintecker**
## Summary

Closes #282 (follow-up to #275, refs #272). The #275 `writable_paths`/Landlock-jail review took **13 rounds / ~30 findings** — most on first-principles properties of the jail's public surface (path containment, glob matching, error-sentinel classes, syscall flag combos) that weren't pinned by tests at implementation time. This PR closes that gap with **property-based tests** so each documented invariant fails fast with a shrunk counter-example when regressed.

**Test-only.** No product code changed. Uses the `pgregory.net/rapid` dev dep that's already in `go.mod` (same library as `tail_buffer_property_test.go`) — **no new dependencies**. Generalizes the example-based table that landed in `88463c6`.

## What's covered

- **`agent/exec/jail_property_test.go`** (cross-platform): escape-vs-malformed sentinel partition of `ValidateWritablePaths`/`validateGlobEntry` (`errors.Is(…, ErrPathEscape)`); every supported doublestar/literal shape validates; `isWindowsAbsolute` vs its spec; `isSubpathOf` containment (reflexive / descendant / prefix-but-not-child); contained-relative vs absolute-escape `working_dir` resolution.
- **`agent/exec/jail_linux_test.go`** (`//go:build linux`): `OpenForWrite` (literal-leaf allow; absolute / parent-escape / symlink-escape refusal; **EACCES-vs-escape** classification); `SafeMkdirAll` (tree creation, idempotent **EEXIST** tolerance, concurrent same-path under `-race`, **symlink-at-any-depth** refusal); `SafeRemove` (remove, **ENOENT-not-escape**, empty-leaf guard, symlink-parent refusal); `landlockDirForGlob` shape + **never-escapes-anchor** invariant.
- **`agent/exec/jail_other_test.go`** (`//go:build !linux`): `RunJailExec` hard-error and `SafeMkdirAll`/`SafeRemove` fail-closed, plus an input-independent refusal property. (Verified via `GOOS=darwin GOARCH=amd64 go test -c`.)
- **`pipeline/handlers/codergen_jail_property_test.go`** (cross-platform): `matchOneGlob`/`matchWritablePath` semantics + union contract; `relPathForJail` containment incl. the bare-`..name` allow case; `configureJail` **G1/G2** refuse-to-start gates (no env hooks wired on refusal); the refuse-backend matrix; and the `WriteOpener`/`Remover` closures (allow / `ErrPathNotAllowed` / `ErrPathEscape` classes, **mkdir-after-glob ordering**) gated on Landlock availability.

## Verification

- `go build ./...` ✓ · `GOOS=darwin GOARCH=amd64 go build ./...` ✓ · `go vet` ✓ · `gofmt` clean
- `go test ./agent/exec/ ./pipeline/handlers/ -race` ✓ · `go test ./... -short` ✓
- `dippin doctor` on the three core pipelines: **Grade A** (unchanged — no engine/.dip edits)
- **Teeth proven:** deliberately mutating the brace-error sentinel and `matchOneGlob`'s descendant branch each fail the corresponding property (rapid shrank to `"a{"` and `matchOneGlob("a/a","a/**")`); both reverted.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---
<a href="https://app.blacksmith.sh/2389-research/codesmith/tracker/pr/293"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-dark-v2.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-light-v2.svg">View with Codesmith</picture></a> <a href="https://backend.blacksmith.sh/track/enable-autofix?expires=1783113315&installation_id=131176874&pr_number=293&repository=2389-research%2Ftracker&return_to=https%3A%2F%2Fgithub.com%2F2389-research%2Ftracker%2Fpull%2F293&signature=c4501720f9a2f6901f2cf0e4928ff7a85f881fe23ff251c0172b4451d558a468"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-light.svg">Autofix with Codesmith</picture></a>
<sup>Need help on this PR? Tag <code>/codesmith</code> with what you need. Autofix is disabled.</sup>

## Summary by CodeRabbit

* **Tests**
  * Expanded test coverage with property-based testing across all platforms (Linux and non-Linux) to ensure system reliability and robustness.
  * Added comprehensive invariant tests for path validation and jail escape prevention logic to catch edge cases.
  * Implemented randomized input testing strategies to strengthen coverage of boundary conditions and degenerate scenarios.

### CODE REVIEW

File: agent/exec/jail_property_test.go [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+	})
+}
+
+// TestIsSubpathOf_Rapid pins the three containment invariants: reflexivity, any
+// literal descendant is contained, and a sibling sharing a string prefix but
+// not a path-component boundary is NOT contained (the bug class isSubpathOf's
+// separator guard exists to prevent).
+func TestIsSubpathOf_Rapid(t *testing.T) {
+	rapid.Check(t, func(t *rapid.T) {
+		parent := "/" + strings.Join(genLiteralSegments(t, "p", 1, 4), "/")
```
[the remark this answers is not in the document]
Comment (clintecker) [reply agent/exec/jail_property_test.go]: Valid — fixed in ccd0434. `isSubpathOf` compares using `string(filepath.Separator)`, so I now build `parent`/`child` with `filepath.Join` (native separators) instead of hard-coded `/`. The descendant case no longer mismatches on Windows. Note the glob generators elsewhere keep `/` deliberately — writable_paths globs are forward-slash by spec regardless of OS. (For context, `agent/exec` is `//go:build !windows` via local.go so it doesn't currently build on Windows, but the test is now correct on its own terms rather than relying on that.)

File: pipeline/handlers/codergen_jail_property_test.go [hunk 2]
```diff
@@ -0,0 +1,413 @@
...
+		}
+		if enabled {
+			t.Fatalf("configureJail enabled=true on G2 refusal for %q", backend)
+		}
+		if !strings.Contains(err.Error(), backend) {
+			t.Fatalf("configureJail err %v does not name backend %q", err, backend)
+		}
+		if env.CommandWrapper != nil || env.WriteOpener != nil {
+			t.Fatalf("env hooks wired despite G2 refusal")
+		}
```
[the remark this answers is not in the document]
Comment (clintecker) [reply pipeline/handlers/codergen_jail_property_test.go]: Good catch — fixed in ccd0434. The G2 "no hooks wired" assertion now includes `env.Remover` alongside `CommandWrapper`/`WriteOpener`, matching the G1 test. A regression that wired `Remover` on a non-native refusal would now fail here.

File: pipeline/handlers/codergen_jail_property_test.go [hunk 3]
```diff
@@ -0,0 +1,413 @@
...
+// --- relPathForJail ---
+
+// TestRelPathForJail_ContainedRoundTrips_Rapid: for any literal relative path,
+// relPathForJail(anchor, anchor/rel) returns rel with no error.
+func TestRelPathForJail_ContainedRoundTrips_Rapid(t *testing.T) {
+	rapid.Check(t, func(t *rapid.T) {
+		anchor := "/" + drawPath(t, "anchor")
+		rel := drawPath(t, "rel")
+		abs := filepath.Join(anchor, rel)
+		got, err := relPathForJail(anchor, abs)
```
[the remark this answers is not in the document]
Comment (clintecker) [reply pipeline/handlers/codergen_jail_property_test.go:181]: Valid — fixed in aacac49. `relPathForJail` returns `filepath.Rel`'s OS-separator output, so the round-trip comparison now normalizes with `filepath.ToSlash(got)` before checking against the forward-slash-generated `rel` (no-op on Linux/darwin, correct on Windows). Same separator-portability class as the earlier `isSubpathOf` fix. The sibling `AllowsDotDotPrefixName` test uses a single-component `..foo` (no separator) and the escape test uses `errors.Is`, so neither needed the change.
